gc.c

```diff
@@ -5198,21 +5212,30 @@ gc_compact_finish(rb_objspace_t *objspace, rb_size_pool_t *pool, rb_heap_t *heap
      * then revert any moved objects that made it to the stack. */
     check_stack_for_moved(objspace);
 
-    gc_update_references(objspace);
-    objspace->profile.compact_count++;
-
     for (int i = 0; i < SIZE_POOL_COUNT; i++) {
         rb_size_pool_t *size_pool = &size_pools[i];
-        rb_heap_t *heap = SIZE_POOL_EDEN_HEAP(size_pool);
-        heap->compact_cursor = NULL;
-        heap->compact_cursor_index = 0;
+        gc_update_references_heap(objspace, size_pool, SIZE_POOL_EDEN_HEAP(size_pool));
     }
 
+    gc_update_references_rest(objspace);
+}
+
+static void
+gc_compact_finish(rb_objspace_t *objspace, rb_size_pool_t *pool, rb_heap_t *heap) {
```

Comment:
IMO it would be nice if the call to `gc_compact_finish_pool` was separate from the call to `gc_compact_finish` (rather than `gc_compact_finish` implicitly calling `gc_compact_finish_pool`).